File: sidecar/agent/mcp.py
"""Model Context Protocol (MCP) client integration.

zWork can use stdio MCP servers listed in ``~/.zwork/mcp.json`` using the
Claude Desktop config shape. Each server tool is exposed to the model as
``mcp__<server>__<tool>`` and routed back through the manager.
"""

from __future__ import annotations

import asyncio
import json
import logging
import os
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Optional

from .home import zwork_home

log = logging.getLogger(__name__)


MCP_CONFIG_FILENAME = "mcp.json"
TOOL_PREFIX = "mcp__"


@dataclass
class MCPServerSpec:
    name: str
    command: str
    args: list[str] = field(default_factory=list)
    env: dict[str, str] = field(default_factory=dict)
    enabled: bool = True


def mcp_config_path() -> Path:
    """Return the path to the MCP server configuration JSON file."""
    return zwork_home() / MCP_CONFIG_FILENAME


def load_config(path: Optional[Path] = None) -> list[MCPServerSpec]:
    p = path or mcp_config_path()
    if not p.exists():
        return []
    try:
        raw = json.loads(p.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError) as e:
        log.warning("mcp.json could not be parsed: %s", e)
        return []
    servers = raw.get("mcpServers") if isinstance(raw, dict) else None
    if not isinstance(servers, dict):
        return []
    out: list[MCPServerSpec] = []
    for name, entry in servers.items():
        if not isinstance(name, str) or not name:
            continue
        if not isinstance(entry, dict):
            continue
        command = entry.get("command")
        if not isinstance(command, str) or not command:
            log.warning("mcp.json: server %r has no command, skipping", name)
            continue
        args_raw = entry.get("args") or []
        env_raw = entry.get("env") or {}
        args = [str(a) for a in args_raw] if isinstance(args_raw, list) else []
        env = (
            {str(k): str(v) for k, v in env_raw.items()}
            if isinstance(env_raw, dict)
            else {}
        )
        enabled = bool(entry.get("enabled", True))
        out.append(
            MCPServerSpec(
                name=name, command=command, args=args, env=env, enabled=enabled
            )
        )
    return out


def prefixed_tool_name(server: str, tool: str) -> str:
    return f"{TOOL_PREFIX}{server}__{tool}"


def split_tool_name(prefixed: str) -> Optional[tuple[str, str]]:
    if not prefixed.startswith(TOOL_PREFIX):
        return None
    rest = prefixed[len(TOOL_PREFIX) :]
    sep = rest.find("__")
    if sep <= 0 or sep >= len(rest) - 2:
        return None
    return rest[:sep], rest[sep + 2 :]


class _Session:
    """Own one MCP stdio session and serialize calls through a queue."""

    def __init__(self, spec: MCPServerSpec) -> None:
        self.spec = spec
        self.tools: list[dict] = []
        self.error: Optional[str] = None
        self.ready = asyncio.Event()
        self._queue: asyncio.Queue = asyncio.Queue()
        self._task: Optional[asyncio.Task] = None

    async def start(self) -> None:
        self._task = asyncio.create_task(self._run(), name=f"mcp-{self.spec.name}")

    async def stop(self) -> None:
        if self._task and not self._task.done():
            await self._queue.put(("__stop__", None, None))
            try:
                await asyncio.wait_for(self._task, timeout=5.0)
            except asyncio.TimeoutError:
                self._task.cancel()
            except Exception:
                pass

    async def call_tool(self, tool_name: str, args: dict) -> dict:
        if not self.ready.is_set():
            await asyncio.wait_for(self.ready.wait(), timeout=10.0)
        if self.error:
            return {"isError": True, "content": [{"type": "text", "text": self.error}]}
        future: asyncio.Future = asyncio.get_running_loop().create_future()
        await self._queue.put(("__call__", tool_name, (args, future)))
        return await asyncio.wait_for(future, timeout=120.0)

    async def _run(self) -> None:
        # Late imports keep zWork bootable when the optional dependency is not
        # installed and no MCP server is configured.
        from mcp import ClientSession, StdioServerParameters
        from mcp.client.stdio import stdio_client

        env = dict(os.environ)
        env.update(self.spec.env)
        params = StdioServerParameters(
            command=self.spec.command, args=self.spec.args, env=env
        )
        try:
            async with stdio_client(params) as (read, write):
                async with ClientSession(read, write) as session:
                    await session.initialize()
                    listed = await session.list_tools()
                    self.tools = [
                        {
                            "name": t.name,
                            "description": t.description or "",
                            "input_schema": t.inputSchema or {"type": "object"},
                        }
                        for t in listed.tools
                    ]
                    self.ready.set()
                    while True:
                        kind, name, payload = await self._queue.get()
                        if kind == "__stop__":
                            return
                        if kind != "__call__":
                            continue
                        args, future = payload
                        try:
                            result = await session.call_tool(name, arguments=args)
                            future.set_result(_serialize_call_result(result))
                        except Exception as e:
                            future.set_result(
                                {
                                    "isError": True,
                                    "content": [
                                        {
                                            "type": "text",
                                            "text": f"{type(e).__name__}: {e}",
                                        }
                                    ],
                                }
                            )
        except Exception as e:
            self.error = f"failed to start MCP server {self.spec.name!r}: {e}"
            log.warning(self.error)
            self.ready.set()


def _serialize_call_result(result: Any) -> dict:
    content_objs = getattr(result, "content", None) or []
    content: list[dict] = []
    for c in content_objs:
        kind = getattr(c, "type", None)
        if kind == "text":
            content.append({"type": "text", "text": getattr(c, "text", "")})
        else:
            content.append({"type": kind or "unknown", "value": str(c)})
    return {"isError": bool(getattr(result, "isError", False)), "content": content}


class MCPManager:
    def __init__(self) -> None:
        self._sessions: dict[str, _Session] = {}

    async def start(self, specs: Optional[list[MCPServerSpec]] = None) -> None:
        if specs is None:
            specs = load_config()
        for spec in specs:
            if not spec.enabled or spec.name in self._sessions:
                continue
            sess = _Session(spec)
            self._sessions[spec.name] = sess
            await sess.start()

    async def stop(self) -> None:
        await asyncio.gather(
            *(s.stop() for s in self._sessions.values()), return_exceptions=True
        )
        self._sessions.clear()

    def status(self) -> list[dict]:
        rows: list[dict] = []
        for name, sess in self._sessions.items():
            rows.append(
                {
                    "name": name,
                    "command": sess.spec.command,
                    "args": list(sess.spec.args),
                    "ready": sess.ready.is_set() and not sess.error,
                    "error": sess.error,
                    "tool_count": len(sess.tools),
                }
            )
        return rows

    def all_tool_schemas(self) -> list[dict]:
        out: list[dict] = []
        for name, sess in self._sessions.items():
            if sess.error or not sess.ready.is_set():
                continue
            for t in sess.tools:
                out.append(
                    {
                        "name": prefixed_tool_name(name, t["name"]),
                        "description": t["description"]
                        or f"MCP tool {t['name']} on {name}",
                        "parameters": t["input_schema"],
                    }
                )
        return out

    async def call_tool(self, prefixed_name: str, args: dict) -> dict:
        parts = split_tool_name(prefixed_name)
        if not parts:
            return {
                "isError": True,
                "content": [
                    {"type": "text", "text": f"not an MCP tool name: {prefixed_name}"}
                ],
            }
        server, tool = parts
        sess = self._sessions.get(server)
        if not sess:
            return {
                "isError": True,
                "content": [{"type": "text", "text": f"unknown MCP server: {server}"}],
            }
        return await sess.call_tool(tool, args)


_manager: Optional[MCPManager] = None


def get_manager() -> MCPManager:
    global _manager
    if _manager is None:
        _manager = MCPManager()
    return _manager

File: sidecar-rust/src/mcp.rs
//! Model Context Protocol (MCP) runtime.
//!
//! Reads configured stdio MCP servers from `~/.zwork/mcp.json` (Claude-Desktop
//! shape: `{"mcpServers": {name: {command, args, env}}}}`), speaks JSON-RPC
//! 2.0 over the child's stdio to list their tools, and exposes them to the
//! agent as `mcp__<server>__<tool>` tools.
//!
//! Each operation spawns the server fresh (initialize → list/call → drop).
//! This is simpler and more robust than managing long-lived sessions: a
//! crashing or hanging server can never wedge the registry, and there's no
//! reconnection logic to get right.

use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::collections::HashMap;
use std::process::Stdio;
use std::time::Duration;
use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader};
use tokio::process::{Child, Command};
use tokio::sync::Mutex;
use tokio::time::timeout;

pub const TOOL_PREFIX: &str = "mcp__";

/// One configured MCP server.
#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct McpServerSpec {
    pub name: String,
    pub command: String,
    #[serde(default)]
    pub args: Vec<String>,
    #[serde(default)]
    pub env: HashMap<String, String>,
    #[serde(default = "default_true")]
    pub enabled: bool,
}

fn default_true() -> bool {
    true
}

/// Path to the MCP config file (`~/.zwork/mcp.json`).
fn config_path() -> std::path::PathBuf {
    crate::paths::home_dir().join("mcp.json")
}

/// Load all enabled server specs from the config file. Returns `[]` if the
/// file is missing or malformed.
pub fn load_config() -> Vec<McpServerSpec> {
    let raw = match std::fs::read_to_string(config_path()) {
        Ok(s) => s,
        Err(_) => return Vec::new(),
    };
    let parsed: Value = match serde_json::from_str(&raw) {
        Ok(v) => v,
        Err(_) => return Vec::new(),
    };
    let servers = parsed.get("mcpServers").and_then(|s| s.as_object());
    let servers = match servers {
        Some(s) => s,
        None => return Vec::new(),
    };
    let mut out = Vec::new();
    for (name, entry) in servers {
        let command = entry.get("command").and_then(|c| c.as_str()).unwrap_or("");
        if command.is_empty() {
            continue;
        }
        let args: Vec<String> = entry.get("args")
            .and_then(|a| a.as_array())
            .map(|a| a.iter().filter_map(|v| v.as_str().map(|s| s.to_string())).collect())
            .unwrap_or_default();
        let env: HashMap<String, String> = entry.get("env")
            .and_then(|e| e.as_object())
            .map(|m| m.iter().filter_map(|(k, v)| v.as_str().map(|s| (k.clone(), s.to_string()))).collect())
            .unwrap_or_default();
        let enabled = entry.get("enabled").and_then(|e| e.as_bool()).unwrap_or(true);
        out.push(McpServerSpec {
            name: name.clone(),
            command: command.to_string(),
            args,
            env,
            enabled,
        });
    }
    out
}

/// A live connection to one MCP server, owning the child process.
struct Connection {
    child: Child,
    stdin: tokio::sync::Mutex<tokio::process::ChildStdin>,
    stdout: tokio::sync::Mutex<BufReader<tokio::process::ChildStdout>>,
}

impl Connection {
    /// Spawn a server and perform the MCP initialize handshake. Returns `Err`
    /// if the process won't start or the handshake times out (10s).
    async fn connect(spec: &McpServerSpec) -> Result<Self, String> {
        let mut cmd = Command::new(&spec.command);
        cmd.args(&spec.args);
        // Start with the current env, then layer the server's env on top.
        for (k, v) in &spec.env {
            cmd.env(k, v);
        }
        cmd.stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped());
        #[cfg(unix)]
        {
            cmd.process_group(0);
        }
        let mut child = cmd.spawn().map_err(|e| format!("Failed to spawn '{}': {}", spec.command, e))?;
        // Drain stderr so it can't fill the OS pipe buffer and deadlock the server.
        if let Some(stderr) = child.stderr.take() {
            let server_name = spec.name.clone();
            tokio::spawn(async move {
                let mut reader = BufReader::new(stderr);
                let mut buf = String::new();
                use tokio::io::AsyncBufReadExt;
                loop {
                    buf.clear();
                    match reader.read_line(&mut buf).await {
                        Ok(0) | Err(_) => break,
                        Ok(_) => {
                            tracing::debug!("mcp[{}]: {}", server_name, buf.trim());
                        }
                    }
                }
            });
        }
        let stdin = child.stdin.take().ok_or("no stdin")?;
        let stdout = child.stdout.take().ok_or("no stdout")?;
        let conn = Connection {
            child,
            stdin: Mutex::new(stdin),
            stdout: Mutex::new(BufReader::new(stdout)),
        };
        // Initialize handshake.
        let init = json!({
            "jsonrpc": "2.0", "id": 0, "method": "initialize",
            "params": {
                "protocolVersion": "2024-11-05",
                "capabilities": {},
                "clientInfo": {"name": "zWork", "version": env!("CARGO_PKG_VERSION")}
            }
        });
        let _ = conn.send_and_receive(&init, 0).await?;
        // Notify initialized.
        conn.notify("notifications/initialized").await?;
        Ok(conn)
    }

    async fn notify(&self, method: &str) -> Result<(), String> {
        let notif = json!({ "jsonrpc": "2.0", "method": method });
        let line = serde_json::to_string(&notif).map_err(|e| e.to_string())?;
        let mut stdin = self.stdin.lock().await;
        stdin.write_all(line.as_bytes()).await.map_err(|e| e.to_string())?;
        stdin.write_all(b"\n").await.map_err(|e| e.to_string())?;
        stdin.flush().await.map_err(|e| e.to_string())?;
        Ok(())
    }

    /// Send a request and read the matching response (matched by id). Skips
    /// notifications and log messages until the response arrives. Times out
    /// after `secs` seconds.
    async fn send_and_receive(&self, request: &Value, secs: u64) -> Result<Value, String> {
        let id = request.get("id").cloned().unwrap_or(json!(0));
        let line = serde_json::to_string(request).map_err(|e| e.to_string())?;
        {
            let mut stdin = self.stdin.lock().await;
            stdin.write_all(line.as_bytes()).await.map_err(|e| e.to_string())?;
            stdin.write_all(b"\n").await.map_err(|e| e.to_string())?;
            stdin.flush().await.map_err(|e| e.to_string())?;
        }
        let mut stdout = self.stdout.lock().await;
        let mut buf = String::new();
        let result = timeout(Duration::from_secs(secs), async {
            loop {
                buf.clear();
                let n = stdout.read_line(&mut buf).await.map_err(|e| e.to_string())?;
                if n == 0 {
                    return Err("MCP server closed stdout".to_string());
                }
                let trimmed = buf.trim();
                if trimmed.is_empty() {
                    continue;
                }
                let msg: Value = match serde_json::from_str(trimmed) {
                    Ok(v) => v,
                    Err(_) => continue, // skip non-JSON lines
                };
                // Match the response by id; ignore notifications (no id).
                if msg.get("id") == Some(&id) {
                    if let Some(err) = msg.get("error") {
                        return Err(format!("MCP error: {}", err));
                    }
                    return Ok(msg.get("result").cloned().unwrap_or(Value::Null));
                }
            }
        })
        .await;
        match result {
            Ok(r) => r,
            Err(_) => Err("MCP request timed out".to_string()),
        }
    }
}

impl Drop for Connection {
    fn drop(&mut self) {
        // Best-effort kill on drop so we never leak a server process.
        let _ = self.child.start_kill();
    }
}

/// List the tools of a single server. Spawns, lists, drops.
async fn list_server_tools(spec: &McpServerSpec) -> (Vec<Value>, Option<String>) {
    let conn = match Connection::connect(spec).await {
        Ok(c) => c,
        Err(e) => return (Vec::new(), Some(e)),
    };
    let req = json!({"jsonrpc":"2.0","id":1,"method":"tools/list"});
    let result = match conn.send_and_receive(&req, 10).await {
        Ok(r) => r,
        Err(e) => return (Vec::new(), Some(e)),
    };
    let tools = result.get("tools").and_then(|t| t.as_array()).cloned().unwrap_or_default();
    (tools, None)
}

/// Build the `mcp__<server>__<tool>` schemas the agent advertises.
pub fn all_tool_schemas() -> Vec<Value> {
    let mut schemas = Vec::new();
    for spec in load_config() {
        if !spec.enabled {
            continue;
        }
        // Use a blocking runtime spawn since list is async but callers are sync.
        let server_name = spec.name.clone();
        let (tools, _err) = block_on_sync(list_server_tools_inner(spec.clone()));
        for t in tools {
            let tool_name = t.get("name").and_then(|n| n.as_str()).unwrap_or("");
            if tool_name.is_empty() {
                continue;
            }
            let prefixed = format!("{}{}__{}", TOOL_PREFIX, server_name, tool_name);
            schemas.push(json!({
                "name": prefixed,
                "description": t.get("description").cloned().unwrap_or_else(|| json!("MCP tool")),
                "parameters": t.get("inputSchema").cloned().unwrap_or_else(|| json!({"type":"object","properties":{}})),
            }));
        }
    }
    schemas
}

async fn list_server_tools_inner(spec: McpServerSpec) -> (Vec<Value>, Option<String>) {
    list_server_tools(&spec).await
}

/// The live status of each configured server, for the `/api/mcp/servers` UI.
pub fn server_status() -> Vec<Value> {
    let specs = load_config();
    let mut out = Vec::new();
    for spec in specs {
        let (tool_count, ready, error) = if spec.enabled {
            let (tools, err) = block_on_sync(list_server_tools_inner(spec.clone()));
            let ready = err.is_none();
            (tools.len(), ready, err)
        } else {
            (0, false, Some("disabled".to_string()))
        };
        out.push(json!({
            "name": spec.name,
            "command": spec.command,
            "args": spec.args,
            "connected": ready,
            "ready": ready,
            "tool_count": tool_count,
            "error": error.unwrap_or_default(),
        }));
    }
    out
}

/// Execute a `mcp__<server>__<tool>` call. Spawns the server, sends
/// `tools/call`, returns the result shaped like a tool result
/// (`{isError, content:[{type,text}]}`).
pub async fn call_tool(prefixed_name: &str, params: Value) -> Value {
    let parts = match split_tool_name(prefixed_name) {
        Some(p) => p,
        None => return json!({
            "isError": true,
            "content": [{"type":"text","text": format!("not an MCP tool: {}", prefixed_name)}]
        }),
    };
    let (server_name, tool_name) = parts;
    let spec = match load_config().into_iter().find(|s| s.name == server_name && s.enabled) {
        Some(s) => s,
        None => return json!({
            "isError": true,
            "content": [{"type":"text","text": format!("MCP server '{}' not configured", server_name)}]
        }),
    };
    match call_server_tool(spec, &tool_name, params).await {
        Ok(v) => v,
        Err(e) => json!({"isError": true, "content": [{"type":"text","text": format!("MCP {}: {}", server_name, e)}]}),
    }
}

async fn call_server_tool(spec: McpServerSpec, tool_name: &str, params: Value) -> Result<Value, String> {
    let conn = Connection::connect(&spec).await?;
    let req = json!({
        "jsonrpc":"2.0","id":1,"method":"tools/call",
        "params": {"name": tool_name, "arguments": params}
    });
    let result = conn.send_and_receive(&req, 120).await?;
    // The MCP tools/call result already has the {isError, content[]} shape —
    // return it as-is.
    Ok(result)
}

/// Split `mcp__<server>__<tool>` into `(server, tool)`. Returns `None` if the
/// name isn't a valid MCP tool name.
pub fn split_tool_name(prefixed: &str) -> Option<(String, String)> {
    let rest = prefixed.strip_prefix(TOOL_PREFIX)?;
    let sep = rest.find("__")?;
    if sep == 0 || sep >= rest.len() - 2 {
        return None;
    }
    Some((rest[..sep].to_string(), rest[sep + 2..].to_string()))
}

// ── Runtime plumbing ─────────────────────────────────────────────────────────
// `all_tool_schemas` / `server_status` are called from sync contexts (the
// schema builder, the route handlers). They need a Tokio runtime to `.await`
// the async server I/O. The route handlers already run on the server's async
// runtime, so we reach for its current handle. `call_tool` is async.

static HANDLE_FALLBACK: std::sync::OnceLock<tokio::runtime::Runtime> = std::sync::OnceLock::new();

/// A runtime handle we can block on. Prefers the currently-running Tokio
/// runtime; if there isn't one (shouldn't happen, but defend against it),
/// spins up a dedicated one.
fn runtime_handle() -> tokio::runtime::Handle {
    if let Ok(h) = tokio::runtime::Handle::try_current() {
        return h;
    }
    let rt = HANDLE_FALLBACK.get_or_init(|| {
        tokio::runtime::Builder::new_multi_thread()
            .enable_all()
            .thread_name("mcp-runtime")
            .build()
            .expect("failed to build MCP runtime")
    });
    rt.handle().clone()
}

/// Block on an async future from a sync context, using `block_in_place` on the
/// current multi-thread runtime worker.
fn block_on_sync<T>(fut: impl std::future::Future<Output = T> + Send + 'static) -> T
where
    T: Send + 'static,
{
    let handle = runtime_handle();
    let join = handle.spawn(fut);
    tokio::task::block_in_place(|| handle.block_on(async { join.await.expect("MCP task panicked") }))
}


